Source/PeleLMRegrid.cpp

```diff
@@ -233,6 +481,67 @@ void PeleLM::ClearLevel(int lev) {
    }
 #endif
    m_extSource[lev]->clear();
+
+   m_costs[lev].reset();
+   m_loadBalanceEff[lev] = -1.0;
+}
+
+void PeleLM::computeCosts(int a_lev, LayoutData<Real> &a_costs, int a_costMethod)
+{
+    if (a_costMethod == LoadBalanceCost::Ncell ) {
+        for (MFIter mfi(a_costs, false); mfi.isValid(); ++mfi)
+        {
+            a_costs[mfi] = mfi.validbox().numPts();
+        }
+    } else if (a_costMethod == LoadBalanceCost::ChemFunctCallAvg ) {
+        MultiFab costMF(a_costs.boxArray(), a_costs.DistributionMap(), 1, 0);
+        fillpatch_chemFunctCall(a_lev, m_cur_time, costMF, 0);
+        for (MFIter mfi(costMF, false); mfi.isValid(); ++mfi)
+        {
+            a_costs[mfi] = costMF[mfi].sum<RunOn::Device>(mfi.validbox(),0) / mfi.validbox().numPts();
+        }
+    } else if (a_costMethod == LoadBalanceCost::ChemFunctCallMax ) {
+        MultiFab costMF(a_costs.boxArray(), a_costs.DistributionMap(), 1, 0);
+        fillpatch_chemFunctCall(a_lev, m_cur_time, costMF, 0);
+        for (MFIter mfi(costMF, false); mfi.isValid(); ++mfi)
+        {
+            a_costs[mfi] = costMF[mfi].max<RunOn::Device>(mfi.validbox(),0);
+        }
+    } else if (a_costMethod == LoadBalanceCost::ChemFunctCallSum ) {
+        MultiFab costMF(a_costs.boxArray(), a_costs.DistributionMap(), 1, 0);
+        fillpatch_chemFunctCall(a_lev, m_cur_time, costMF, 0);
+        for (MFIter mfi(costMF, false); mfi.isValid(); ++mfi)
+        {
+            a_costs[mfi] = costMF[mfi].sum<RunOn::Device>(mfi.validbox(),0);
+        }
+    } else if (a_costMethod == LoadBalanceCost::UserDefinedDerivedAvg) {
+        MultiFab costMF(a_costs.boxArray(), a_costs.DistributionMap(), 1, 0);
+        costMF.setVal(0.0);
+        std::unique_ptr<MultiFab> mf;
+        mf = derive("derUserDefined", m_cur_time, a_lev, 0);
+        costMF.ParallelCopy(*mf,0,0,1);
+        for (MFIter mfi(costMF, false); mfi.isValid(); ++mfi)
+        {
+            a_costs[mfi] = costMF[mfi].sum<RunOn::Device>(mfi.validbox(),0) / mfi.validbox().numPts();
+        }
+    } else if (a_costMethod == LoadBalanceCost::UserDefinedDerivedSum) {
+        MultiFab costMF(a_costs.boxArray(), a_costs.DistributionMap(), 1, 0);
+        costMF.setVal(0.0);
+        std::unique_ptr<MultiFab> mf;
+        mf = derive("derUserDefined", m_cur_time, a_lev, 0);
```

Comment:
There's an edge case here where multiple variables are specified in "derUserDefined" and this requires that the one used for load balancing is the first. I guess any user that is using this functionality should be advanced enough to figure this out though.

Reply:
Good point, I added a comment to the control section.